File: 127-word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        patternMap = defaultdict(list)
        for word in wordList:
            for i in range(len(word)):
                pattern = word[:i] + '*' + word[i+1:]
                patternMap[pattern].append(word)
        queue = deque()
        queue.append((beginWord,1))
        visited = set()
        level = 1
        while queue:
            currWord,level = queue.popleft()
            visited.add(currWord)
            for i in range(len(word)):
                pattern = currWord[:i] + '*' + currWord[i+1:]
                for adjWord in patternMap[pattern]:
                    if adjWord == endWord:
                        return level + 1
                    if adjWord not in visited:
                        queue.append((adjWord,level + 1))
        return 0

File: 127-word-ladder/word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>(wordList);
        if (!set.contains(endWord)) return 0;
        Queue<String> queue = new LinkedList<>();
        queue.add(beginWord);
        int level = 1;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            for (int k = 0; k < levelSize; k++) {
                char[] currentWord = queue.poll().toCharArray();
                for (int i = 0; i < currentWord.length; i++) {
                    char originalChar = currentWord[i];
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == originalChar) continue;
                        currentWord[i] = c;
                        String newWord = new String(currentWord);
                        if (set.contains(newWord)) {
                            if (newWord.equals(endWord)) {
                                return level + 1;
                            }
                            queue.add(newWord);
                            set.remove(newWord);
                        }
                    }
                    currentWord[i] = originalChar;
                }
            }
            level++;
        }
        return 0;
    }
}
